### ISSUE TITLE
A virtual keyboard on a floating window 

### ISSUE THREAD

**Issue by Alessandro1970** [dirkwhoffmann/virtualc64#303]
A virtual keyboard on a floating window, pressing the respective keys the layout change untill pressed again, so the users can select all the desired keys.
By pressing on the virtual keyboard, with the mouse, the respective function-key the keyboard layout changes as shown (three options).
A floating windows for me is the best solution for this emulator so the users can move it on the desired place in the desktop, then point and click on the image and the letter or symbol will appare on the emulator like a real C64 keyboard.

![37402601-88175fb8-278c-11e8-82bc-03fe28b70c49](https://user-images.githubusercontent.com/34916641/37405960-4e8ffaf2-2796-11e8-89e3-d0d0d3905661.jpg)

**Comment by bluecursor** [comment]
I think this is a great addon, specially for users who are using seq editors.

**Comment by puleyo** [comment]
Great !
I know that a friend of mine has a virtual keboard on a c64 emulator on linux.
Can it be added in VirtualC64 ?

<img width="719" alt="schermata 2018-04-21 alle 11 29 58" src="https://user-images.githubusercontent.com/38581054/39082709-b3b79f8c-4557-11e8-8bee-b978ebe488a5.png">

**Comment by Alessandro1970** [comment]
Yes! 👍🏻

**Comment by Ottavius** [comment]
Oh great, I read now about this improvement. It would be great to have a virtual keyboard that can be used with the mouse of the mac; it would be easier to find keys and graphic symbols and it would be immediate to use them with a simple click...

**Comment by dirkwhoffmann** [comment]
I think a nice solution would be to implement a C64 keyboard as a standalone app and add some API to VirtualC64 (AppleScript or whatever is hot these days) that accepts key presses from an external app. Does anybody know if such a keyboard app is already there?

**Comment by Ottavius** [comment]
I think it would be more useful to have a small window showing the commodore's keyboard as an integral part of the emulator; and be able to use the mouse to choose the right button or the graphic character with a simple click, and not an external app. A virtual keyboard as used on commodore emulators in smartphones.

**Comment by puleyo** [comment]
I was thinking of a virtual keyboard like the cmmodore emulator for the old PSP ... it worked well, and fine.

![pspvice-05](https://user-images.githubusercontent.com/38581054/42151833-80c98274-7dde-11e8-8b04-ec0db6348e49.png)

http://perso.numericable.fr/~ckckck/C64_PSPVice.html

**Comment by Alessandro1970** [comment]
Yes the VICE version for the PSP had a great virtual keyboard (point and click).
It worked with graphic characters too.
**Is it possible to import into the VirtualC64 the keyboard designed for the PSP?**

![screenshot001](https://user-images.githubusercontent.com/34916641/42152127-85a107ee-7ddf-11e8-9347-6f3b5e69e3a7.jpg)

https://www.c64-wiki.com/wiki/PSPVice

**Comment by Alessandro1970** [comment]
PSP VICE Source code:

http://svn.akop.org/psp/tags/vice/2.2.15/

**Comment by Ottavius** [comment]
Beautiful keyboard, you can also see the graphic characters, great.

**Comment by dada08ms** [comment]
What splendor this one !

![42151833-80c98274-7dde-11e8-8b04-ec0db6348e49](https://user-images.githubusercontent.com/40881549/42362797-45eb1ca2-80f5-11e8-8079-f61717281248.png)

**Comment by dirkwhoffmann** [comment]
Wouldn't it be nicer to have some kind of a semi-transparent overlay as in this mockup (with a C64 keyboard of course)?
![mockup](https://user-images.githubusercontent.com/12561945/42365825-da52c660-80ff-11e8-879b-261b4359c985.png)

The general problem with overlays, however, is that it's not that easy to implement. It has to be part of the 3D Metal scene, so it's size and position depends on the view point matrices. Right now, I have no glue how the map the 2D coordinate (which I get when the mouse is clicked) to the corresponding part of the 3D scene.

**Comment by Alessandro1970** [comment]
Aesthetically it is superb, but to avoid problems would it be better that the emulator generates a secondary window (external window) isuch that the pointer does not create position conflict and also with the possible emulation of the commodore mouse ?

**Comment by Ottavius** [comment]
Matter of taste.
But as useful, I can not write if something obstructs my vision and transparency does not help me.
On a large monitor I would prefer to put the VirtualC64 on the top left and at the bottom right the keyboard so there are no problems neither of view nor of conflict between the position of the emulator window and the virtual keyboard window.

**Comment by dirkwhoffmann** [comment]
I just clicked together two preliminary visual keyboard prototypes. This is how it could look like:
 
<img width="950" alt="prototype2" src="https://user-images.githubusercontent.com/12561945/43368481-f31d20d0-935d-11e8-8b77-6bebdb7fc4d5.png">

<img width="936" alt="prototype1" src="https://user-images.githubusercontent.com/12561945/43368482-f552e646-935d-11e8-8831-708f0f1da015.png">

From a software engineering point of view, I plan to implement the keyboard as a stand alone app and let it communicate with the emulator via notification center. Having two separate apps has the architectural advantage that I can keep the emulator lean.

**Comment by dirkwhoffmann** [comment]
A couple of weeks ago, I found a keyboard layout schematics on the web that showed the exact key dimensions and the spaces in between. Unfortunately, I cannot find this image any more. If somebody comes across such a picture, please post.

**Comment by Alessandro1970** [comment]
Ok, even if we would have preferred an integrated keyboard with the emulator, which would have been a step forward compared to all the other emulation programs that do not present it, and which in fact represents a real lack because without the possibility to emulate, in all respects, a keyboard the emulation would never be complete.
I have these keyboard scheme layout:
![p4-07l](https://user-images.githubusercontent.com/34916641/43378809-5adc80c4-93c9-11e8-813e-26790ebd8f4d.gif)

...there was an old emulator (WIN XP) for c64 which had a virtual keyboard:
https://github.com/ThKattanek/emu64

![untitled 2](https://user-images.githubusercontent.com/34916641/43379105-0ec15050-93cb-11e8-8de9-1dd116561bf0.jpg)

**Comment by Alessandro1970** [comment]
> external one but can be recalled by an icon on the toolbar

> its own folder with the VirtualC64 itself, a file of documents, a folder to insert the requested ROMs, and the virtual keyboard (a bit like the other emulators), etc.

Ok, good idea, this would be simple for the user who would still see everything as a single application without getting confused between the myriad of apps scattered around the mac.
The icon to call up the toolbar keyboard would be a great solution, both for ease and immediacy of use.

**Comment by dirkwhoffmann** [comment]
Update about my findings regarding inter process communication (IPC)

Using NotificationCenter works, but has the drawback that sandboxed applications are forbidden to communicate data (userInfo must be set to nil). In my prototype, I coped with that by coding the row and column information of the key into the message name (e.g., keyPressed42 for pressing the key in row 4 and column 2).

Note: To get an instant response from the receiving task, it's necessary to send the notification with flag "deliverImmediately" set to true.

    DistributedNotificationCenter.default().postNotificationName(nname, object: nil, userInfo: nil, deliverImmediately: true)

Next, I'll look into AppleScript ...

**Comment by dirkwhoffmann** [comment]
Applescript is zombie technology (theoretically alive, but practically dead). Based on the available options and the discussion in this thread, I'll implement the keyboard inside the emulator and not as an external app.

**Comment by Alessandro1970** [comment]
Thank YOU !

**Comment by dirkwhoffmann** [comment]
Presumably, it'll look like this: 

<img width="684" alt="prototype" src="https://user-images.githubusercontent.com/12561945/43401993-db0ff24c-9411-11e8-9fb1-81f5dfda5514.png">

**Comment by puleyo** [comment]
Beautiful, but doest this keyboard change layout by pressing the button and show the graphic characters instead of the letters?

Under the F1 key should F2 appear
...or does by pressing the control keys the layaout change ?

**Comment by dirkwhoffmann** [comment]
Implemented in V2.4 beta 2: 

http://dirkwhoffmann.de/VirtualC64_2.4beta2.zip

I guess, Puleyo's comment "A virtual keyboard is required for a quick use of keys that are not immediately available" is the key point. It led me to the following design decision: The virtual keyboard opens as a sheet when clicking the corresponding toolbar icon and closes by itself when a key has been hit other than left Shift, right Shift, Shift Lock, or Commodore. When one of these special keys is pressed, they become highlighted, and in the case of the shift keys, the visual appearance of the keyboard changes. The Commodore key has no effect on the visual appearance right now.

<img width="834" alt="virtualkeyboard" src="https://user-images.githubusercontent.com/12561945/43474924-8029def4-94f4-11e8-98ec-887efcc4a364.png">

**Comment by Alessandro1970** [comment]
Wow! This rapid keyboard is truly formidable.
Very happy
The Commodore key is not yet implemented but this is the state of art.

This virtual keyboard is incredibly fast, intuitive, versatile. ...beautiful!
... Poor VICE ... a new king has arisen!
W the VirtualC64 !

**Comment by Ottavius** [comment]
Excellent result, fast and effective virtual keyboard.
Thank you

**Comment by puleyo** [comment]
Great result, I like it very much.
Will it be possible to display the graphic characters when the C= key is pressed?
...maybe in the final version 2.4?

**Comment by dirkwhoffmann** [comment]
2.4 has been released.
In the final version, graphic characters appear when shift or commodore is pressed. 

I kept the "RunStop + Restore" item in the "Keyboard->Press Key" menu. All other items have been removed as it seemed simpler for me to press those keys via the new keyboard. All other items (Home, Clear etc.) are simply marked as hidden, so I can resurrect them immediately if somebody wants them back.
